qiskit/optimization/algorithms/warm_start_qaoa_optimizer.py

```diff
@@ -0,0 +1,318 @@
+
+import numpy as np
+from qiskit import QuantumCircuit
+from qiskit.aqua import AquaError
+from qiskit.aqua.algorithms import QAOA
+from qiskit.circuit import Parameter
+from qiskit.optimization import QuadraticProgram, QiskitOptimizationError
+from qiskit.optimization.algorithms import OptimizationAlgorithm, MinimumEigenOptimizer, \
+    MinimumEigenOptimizationResult, OptimizationResultStatus
+from qiskit.optimization.converters import QuadraticProgramToQubo, QuadraticProgramConverter
+from qiskit.optimization.problems import VarType
+
+
+class BaseAggregator(ABC):
+    """Aggregates results"""
+
+    def aggregate(self, results: List[MinimumEigenOptimizationResult], problem: QuadraticProgram,
+                  qubo_converter: QuadraticProgramToQubo) -> MinimumEigenOptimizationResult:
+        """Aggregates the results."""
+        raise NotImplementedError
+
+
+class MeanAggregator(BaseAggregator):
+    """Aggregates the results by averaging the probability of each sample."""
+
+    def aggregate(self, results: List[MinimumEigenOptimizationResult],
+                  problem: QuadraticProgram, qubo_converter: QuadraticProgramToQubo) \
+            -> MinimumEigenOptimizationResult:
+        """
+        Args:
+            results: List of result objects that need to be combined.
+            problem: A converted problem being solved by the optimizer.
+            qubo_converter: A converter used to convert the initial problem.
+
+        Returns:
+             An aggregated result.
+        """
+
+        # Use a dict for fast solution look-up
+        dict_samples: Dict[str, Tuple[float, float]] = {}
+
+        # Sum up all the probabilities in the results
+        for result in results:
+            for sample in result.samples:
+                if sample[0] in dict_samples:
+                    dict_samples[sample[0]] = (sample[1], dict_samples[sample[0]][1] + sample[2])
+                else:
+                    dict_samples[sample[0]] = (sample[1], sample[2])
+
+        # Divide by the number of results to normalize
+        new_samples = []
+        n_results = len(results)
+        for state in dict_samples:
+            sample = (state, dict_samples[state][0], dict_samples[state][1] / n_results)
+            new_samples.append(sample)
+
+        # todo: this snippet is also used in min eigen optimizer, consider making it as a function
+        new_samples.sort(key=lambda sample: problem.objective.sense.value * sample[1])
+        x = [float(e) for e in new_samples[0][0]]
+        fval = new_samples[0][1]
+
+        # todo: we have to call _interpret to convert to the original representation
+        return MinimumEigenOptimizationResult(x, fval, variables=problem.variables,
+                                              status=OptimizationResultStatus.SUCCESS,
+                                              samples=new_samples)
+
+
+class MixerFactory(ABC):
+    """An abstract factory for creating mixers for warm start QAOA."""
```

Comment:
```suggestion
    """An abstract factory for creating mixers for QAOA."""
```
Could we use this for any QAOA? I.e. suppose I want to try different mixer operators for QAOA, then the `MixerFactory` could be a good means of doing this. I'm not sure what the signature of `create_mixer` should look like. Perhaps `create_mixer(self, **kwargs)`?